File: Programming/Haskell/l13z1.hs
import Control.Monad
import Data.List

hetman :: Int -> [[Int]]
hetman n  = permH [1..n] where
    permH [] = return []
    permH xs = do
        x <- xs
        let xs' = delete x xs
        hs <- permH xs'
        guard(sprawdz x hs 1)
        return (x:hs)

    sprawdz _ [] _ = True
    sprawdz x (y:ys) odl = 
        abs (x-y) /= odl && sprawdz x ys (odl+1)

File: Programming/Prolog/l13z1.pl
hetman(N,X):-
	listaEnum(1,N,L),
	permH(L,X).

listaEnum(N,N,[N]):-!.
listaEnum(L,N,[L|T]):-
	L1 is L+1,
	listaEnum(L1,N,T).

permH([],[]).
permH(L,[H|T]):-
	select(H,L,L1),
	permH(L1,T),
	sprawdz(H,T,1).

sprawdz(_,[],_).
sprawdz(X,[H|T],Odl):-
	not(Odl is abs(X-H)),
	Odl1 is Odl + 1,
	sprawdz(X,T,Odl1).
